src/user/tipsandtricks.rst

```diff
@@ -130,3 +130,25 @@ To verify that the correct platform is being used, run the following commands af
 
     python -c "import platform;print(platform.machine())"  # Should print "x86_64"
     echo "CONDA_SUBDIR: $CONDA_SUBDIR"  # Should print "CONDA_SUBDIR: osx-64"
+
+
+.. _installing_packages_for_GPUs_and_CPUs:
+
+Installing CUDA-enabled packages like TensorFlow and PyTorch 
+============================================================
+
+In conda-forge, some packages are available with GPU support. These packages not only take significantly longer to compile and build, but they also result in rather large binaries that users then download. As an effort to maximize accessibility for users with lower connection and/or storage bandwidth, there is an ongoing effort to limit installing packages compiled for GPUs unnecessarily on CPU-only machines by default. This is accomplished by adding a run dependency, ``__cuda``, that detects if the local machine has a GPU. However, this introduces challenges to users who may prefer to still download and use ``-gpu`` packages even on a non-GPU machine. For example, login nodes on HPCs often do not have GPUs and their compute counterparts with GPUs often do not have internet access. In this case, a user can override the default setting via the environment variable ``CONDA_CUDA_OVERRIDE`` to install GPU packages on the login node to be used later on the compute node. At the time of writing (February 2022), we have concluded this safe default behavior is best for most of conda-forge users, with an easy override option available and documented. Please let us know if you have thoughts on or issues with this.
+
+In order to override the default behavior, a user can set the environment variable ``CONDA_CUDA_OVERRIDE`` like below to install TensorFlow with GPU support even on a machine with CPU only.
+
+.. code-block:: shell
+
+     CONDA_CUDA_OVERRIDE="11.2" conda install tensorflow -c conda-forge
+     # OR
+     CONDA_CUDA_OVERRIDE="11.2" mamba install tensorflow -c conda-forge
+
+.. note::
+   
+   Note that you should select the cudatoolkit version most appropraite for your GPU; currently, we have "10.2", "11.0", "11.1", and "11.2" available.
```

Comment:
This is the tricky part. Newer CUDA driver (say 11.4) can run with older CTK version (say 11.2), but not the other way around. Also, we already have up to CTK 11.6 on conda-forge.

cc: @jakirkham @jaimergp @kkraus14

Reply:
I initially thought his would need to mirror the `cuda_compiler_version` variable, where we now only have the above (10.2 through 11.2). I am not sure though, so any feedback is welcome!

Reply:
To be clear the full text around this issue is:

```
In order to override the default behavior, a user can set the environment variable ``CONDA_CUDA_OVERRIDE`` like below to install TensorFlow with GPU support even on a machine with CPU only.

 .. code-block:: shell
      CONDA_CUDA_OVERRIDE="11.2" conda install tensorflow -c conda-forge
      # OR
      CONDA_CUDA_OVERRIDE="11.2" mamba install tensorflow -c conda-forge
 .. note::

    Note that you should select the cudatoolkit version most appropraite for your GPU; currently, we have "10.2", "11.0", "11.1", and "11.2" available.
```

Reply:
`cuda_compiler_version` only matters for feedstock maintainers; for end users, they see all 11.x (up to 6 currently).

Reply:
They all pull in the same exact thing actually, so not sure if this variable needs to be specified at all like this.

<details>

```

$ CONDA_CUDA_OVERRIDE="10.2" mamba install tensorflow -c conda-forge

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.20.0) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['tensorflow']

conda-forge/linux-64     Using cache
conda-forge/noarch       Using cache

Pinned packages:
  - python 3.9.*


Transaction

  Prefix: /home/ngam/.Mambaforge-Linux-x86_64

  Updating specs:

   - tensorflow
   - ca-certificates
   - certifi
   - openssl


  Package                       Version  Build                   Channel                     Size
───────────────────────────────────────────────────────────────────────────────────────────────────
  Install:
───────────────────────────────────────────────────────────────────────────────────────────────────

  + abseil-cpp               20210324.2  h9c3ff4c_0              conda-forge/linux-64     1010 KB
  + absl-py                      0.15.0  pyhd8ed1ab_0            conda-forge/noarch         98 KB
  + aiohttp                       3.8.1  py39h3811e60_0          conda-forge/linux-64      588 KB
  + aiosignal                     1.2.0  pyhd8ed1ab_0            conda-forge/noarch         12 KB
  + astunparse                    1.6.3  pyhd8ed1ab_0            conda-forge/noarch         15 KB
  + async-timeout                 4.0.2  pyhd8ed1ab_0            conda-forge/noarch          9 KB
  + blinker                         1.4  py_1                    conda-forge/noarch        Cached
  + cached-property               1.5.2  hd8ed1ab_1              conda-forge/noarch          4 KB
  + cached_property               1.5.2  pyha770c72_1            conda-forge/noarch        Cached
  + cachetools                    4.2.4  pyhd8ed1ab_0            conda-forge/noarch         12 KB
  + click                         8.0.3  py39hf3d152e_1          conda-forge/linux-64      147 KB
  + cudatoolkit                  11.6.0  habf752d_10             conda-forge/linux-64      821 MB
  + cudnn                      8.2.1.32  h86fa8c9_0              conda-forge/linux-64      674 MB
  + frozenlist                    1.3.0  py39h3811e60_0          conda-forge/linux-64      161 KB
  + gast                          0.4.0  pyh9f0ad1d_0            conda-forge/noarch         12 KB
  + giflib                        5.2.1  h36c2ea0_2              conda-forge/linux-64      Cached
  + google-auth                  1.35.0  pyh6c4a22f_0            conda-forge/noarch         81 KB
  + google-auth-oauthlib          0.4.6  pyhd8ed1ab_0            conda-forge/noarch         19 KB
  + google-pasta                  0.2.0  pyh8c360ce_0            conda-forge/noarch         42 KB
  + grpc-cpp                     1.41.1  h75e9d12_2              conda-forge/linux-64        4 MB
  + grpcio                       1.41.1  py39hff7568b_1          conda-forge/linux-64        2 MB
  + h5py                          3.1.0  nompi_py39h25020de_100  conda-forge/linux-64        1 MB
  + hdf5                         1.10.6  nompi_h6a2412b_1114     conda-forge/linux-64        3 MB
  + jpeg                             9e  h7f98852_0              conda-forge/linux-64      Cached
  + keras                         2.6.0  pyhd8ed1ab_1            conda-forge/noarch        822 KB
  + keras-preprocessing           1.1.2  pyhd8ed1ab_0            conda-forge/noarch         34 KB
  + libblas                       3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libcblas                      3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libgfortran-ng               11.2.0  h69a702a_12             conda-forge/linux-64       19 KB
  + libgfortran5                 11.2.0  h5c6108e_12             conda-forge/linux-64      Cached
  + liblapack                     3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libopenblas                  0.3.18  pthreads_h8fe5266_0     conda-forge/linux-64      Cached
  + libpng                       1.6.37  h21135ba_2              conda-forge/linux-64      Cached
  + libprotobuf                  3.18.1  h780b84a_0              conda-forge/linux-64        3 MB
  + markdown                      3.3.6  pyhd8ed1ab_0            conda-forge/noarch         67 KB
  + multidict                     6.0.2  py39h3811e60_0          conda-forge/linux-64      156 KB
  + nccl                       2.11.4.1  h5c60f85_2              conda-forge/linux-64      146 MB
  + numpy                        1.19.5  py39hdbf815f_2          conda-forge/linux-64        5 MB
  + oauthlib                      3.2.0  pyhd8ed1ab_0            conda-forge/noarch         90 KB
  + opt_einsum                    3.3.0  pyhd8ed1ab_1            conda-forge/noarch        Cached
  + protobuf                     3.18.1  py39he80948d_0          conda-forge/linux-64      347 KB
  + pyasn1                        0.4.8  py_0                    conda-forge/noarch        Cached
  + pyasn1-modules                0.2.7  py_0                    conda-forge/noarch        Cached
  + pyjwt                         2.3.0  pyhd8ed1ab_1            conda-forge/noarch        Cached
  + python-flatbuffers             1.12  pyhd8ed1ab_1            conda-forge/noarch         19 KB
  + pyu2f                         0.1.5  pyhd8ed1ab_0            conda-forge/noarch        Cached
  + re2                      2021.11.01  h9c3ff4c_0              conda-forge/linux-64      222 KB
  + requests-oauthlib             1.3.1  pyhd8ed1ab_0            conda-forge/noarch         22 KB
  + rsa                             4.8  pyhd8ed1ab_0            conda-forge/noarch        Cached
  + scipy                         1.8.0  py39hee8e79c_1          conda-forge/linux-64       25 MB
  + snappy                        1.1.8  he1b5a44_3              conda-forge/linux-64       32 KB
  + tensorboard                   2.6.0  pyhd8ed1ab_1            conda-forge/noarch          5 MB
  + tensorboard-data-server       0.6.0  py39h95dcef6_1          conda-forge/linux-64        3 MB
  + tensorboard-plugin-wit        1.8.1  pyhd8ed1ab_0            conda-forge/noarch        668 KB
  + tensorflow                    2.6.2  cuda112py39h9333c2f_1   conda-forge/linux-64       29 KB
  + tensorflow-base               2.6.2  cuda112py39he9472f8_1   conda-forge/linux-64      331 MB
  + tensorflow-estimator          2.6.2  cuda112py39h9333c2f_1   conda-forge/linux-64      654 KB
  + termcolor                     1.1.0  py_2                    conda-forge/noarch        Cached
  + typing-extensions           3.7.4.3  0                       conda-forge/noarch          8 KB
  + typing_extensions           3.7.4.3  py_0                    conda-forge/noarch         25 KB
  + werkzeug                      2.0.3  pyhd8ed1ab_1            conda-forge/noarch        306 KB
  + wrapt                        1.12.1  py39h3811e60_3          conda-forge/linux-64       47 KB
  + yarl                          1.7.2  py39h3811e60_1          conda-forge/linux-64      138 KB

  Downgrade:
───────────────────────────────────────────────────────────────────────────────────────────────────

  - six                          1.16.0  pyh6c4a22f_0            installed                       
  + six                          1.15.0  pyh9f0ad1d_0            conda-forge/noarch         14 KB

  Summary:

  Install: 63 packages
  Downgrade: 1 packages

  Total download: 2 GB

───────────────────────────────────────────────────────────────────────────────────────────────────

Confirm changes: [Y/n] 

```

</details>

<details>

```
$ CONDA_CUDA_OVERRIDE="11.2" mamba install tensorflow -c conda-forge

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.20.0) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['tensorflow']

conda-forge/linux-64     Using cache
conda-forge/noarch       Using cache

Pinned packages:
  - python 3.9.*


Transaction

  Prefix: /home/ngam/.Mambaforge-Linux-x86_64

  Updating specs:

   - tensorflow
   - ca-certificates
   - certifi
   - openssl


  Package                       Version  Build                   Channel                     Size
───────────────────────────────────────────────────────────────────────────────────────────────────
  Install:
───────────────────────────────────────────────────────────────────────────────────────────────────

  + abseil-cpp               20210324.2  h9c3ff4c_0              conda-forge/linux-64     1010 KB
  + absl-py                      0.15.0  pyhd8ed1ab_0            conda-forge/noarch         98 KB
  + aiohttp                       3.8.1  py39h3811e60_0          conda-forge/linux-64      588 KB
  + aiosignal                     1.2.0  pyhd8ed1ab_0            conda-forge/noarch         12 KB
  + astunparse                    1.6.3  pyhd8ed1ab_0            conda-forge/noarch         15 KB
  + async-timeout                 4.0.2  pyhd8ed1ab_0            conda-forge/noarch          9 KB
  + blinker                         1.4  py_1                    conda-forge/noarch        Cached
  + cached-property               1.5.2  hd8ed1ab_1              conda-forge/noarch          4 KB
  + cached_property               1.5.2  pyha770c72_1            conda-forge/noarch        Cached
  + cachetools                    4.2.4  pyhd8ed1ab_0            conda-forge/noarch         12 KB
  + click                         8.0.3  py39hf3d152e_1          conda-forge/linux-64      147 KB
  + cudatoolkit                  11.6.0  habf752d_10             conda-forge/linux-64      821 MB
  + cudnn                      8.2.1.32  h86fa8c9_0              conda-forge/linux-64      674 MB
  + frozenlist                    1.3.0  py39h3811e60_0          conda-forge/linux-64      161 KB
  + gast                          0.4.0  pyh9f0ad1d_0            conda-forge/noarch         12 KB
  + giflib                        5.2.1  h36c2ea0_2              conda-forge/linux-64      Cached
  + google-auth                  1.35.0  pyh6c4a22f_0            conda-forge/noarch         81 KB
  + google-auth-oauthlib          0.4.6  pyhd8ed1ab_0            conda-forge/noarch         19 KB
  + google-pasta                  0.2.0  pyh8c360ce_0            conda-forge/noarch         42 KB
  + grpc-cpp                     1.41.1  h75e9d12_2              conda-forge/linux-64        4 MB
  + grpcio                       1.41.1  py39hff7568b_1          conda-forge/linux-64        2 MB
  + h5py                          3.1.0  nompi_py39h25020de_100  conda-forge/linux-64        1 MB
  + hdf5                         1.10.6  nompi_h6a2412b_1114     conda-forge/linux-64        3 MB
  + jpeg                             9e  h7f98852_0              conda-forge/linux-64      Cached
  + keras                         2.6.0  pyhd8ed1ab_1            conda-forge/noarch        822 KB
  + keras-preprocessing           1.1.2  pyhd8ed1ab_0            conda-forge/noarch         34 KB
  + libblas                       3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libcblas                      3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libgfortran-ng               11.2.0  h69a702a_12             conda-forge/linux-64       19 KB
  + libgfortran5                 11.2.0  h5c6108e_12             conda-forge/linux-64      Cached
  + liblapack                     3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libopenblas                  0.3.18  pthreads_h8fe5266_0     conda-forge/linux-64      Cached
  + libpng                       1.6.37  h21135ba_2              conda-forge/linux-64      Cached
  + libprotobuf                  3.18.1  h780b84a_0              conda-forge/linux-64        3 MB
  + markdown                      3.3.6  pyhd8ed1ab_0            conda-forge/noarch         67 KB
  + multidict                     6.0.2  py39h3811e60_0          conda-forge/linux-64      156 KB
  + nccl                       2.11.4.1  h5c60f85_2              conda-forge/linux-64      146 MB
  + numpy                        1.19.5  py39hdbf815f_2          conda-forge/linux-64        5 MB
  + oauthlib                      3.2.0  pyhd8ed1ab_0            conda-forge/noarch         90 KB
  + opt_einsum                    3.3.0  pyhd8ed1ab_1            conda-forge/noarch        Cached
  + protobuf                     3.18.1  py39he80948d_0          conda-forge/linux-64      347 KB
  + pyasn1                        0.4.8  py_0                    conda-forge/noarch        Cached
  + pyasn1-modules                0.2.7  py_0                    conda-forge/noarch        Cached
  + pyjwt                         2.3.0  pyhd8ed1ab_1            conda-forge/noarch        Cached
  + python-flatbuffers             1.12  pyhd8ed1ab_1            conda-forge/noarch         19 KB
  + pyu2f                         0.1.5  pyhd8ed1ab_0            conda-forge/noarch        Cached
  + re2                      2021.11.01  h9c3ff4c_0              conda-forge/linux-64      222 KB
  + requests-oauthlib             1.3.1  pyhd8ed1ab_0            conda-forge/noarch         22 KB
  + rsa                             4.8  pyhd8ed1ab_0            conda-forge/noarch        Cached
  + scipy                         1.8.0  py39hee8e79c_1          conda-forge/linux-64       25 MB
  + snappy                        1.1.8  he1b5a44_3              conda-forge/linux-64       32 KB
  + tensorboard                   2.6.0  pyhd8ed1ab_1            conda-forge/noarch          5 MB
  + tensorboard-data-server       0.6.0  py39h95dcef6_1          conda-forge/linux-64        3 MB
  + tensorboard-plugin-wit        1.8.1  pyhd8ed1ab_0            conda-forge/noarch        668 KB
  + tensorflow                    2.6.2  cuda112py39h9333c2f_1   conda-forge/linux-64       29 KB
  + tensorflow-base               2.6.2  cuda112py39he9472f8_1   conda-forge/linux-64      331 MB
  + tensorflow-estimator          2.6.2  cuda112py39h9333c2f_1   conda-forge/linux-64      654 KB
  + termcolor                     1.1.0  py_2                    conda-forge/noarch        Cached
  + typing-extensions           3.7.4.3  0                       conda-forge/noarch          8 KB
  + typing_extensions           3.7.4.3  py_0                    conda-forge/noarch         25 KB
  + werkzeug                      2.0.3  pyhd8ed1ab_1            conda-forge/noarch        306 KB
  + wrapt                        1.12.1  py39h3811e60_3          conda-forge/linux-64       47 KB
  + yarl                          1.7.2  py39h3811e60_1          conda-forge/linux-64      138 KB

  Downgrade:
───────────────────────────────────────────────────────────────────────────────────────────────────

  - six                          1.16.0  pyh6c4a22f_0            installed                       
  + six                          1.15.0  pyh9f0ad1d_0            conda-forge/noarch         14 KB

  Summary:

  Install: 63 packages
  Downgrade: 1 packages

  Total download: 2 GB

───────────────────────────────────────────────────────────────────────────────────────────────────

Confirm changes: [Y/n] 
```

</details>

<details>

```
$ CONDA_CUDA_OVERRIDE="11.6" mamba install tensorflow -c conda-forge

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.20.0) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['tensorflow']

conda-forge/linux-64     Using cache
conda-forge/noarch       Using cache

Pinned packages:
  - python 3.9.*


Transaction

  Prefix: /home/ngam/.Mambaforge-Linux-x86_64

  Updating specs:

   - tensorflow
   - ca-certificates
   - certifi
   - openssl


  Package                       Version  Build                   Channel                     Size
───────────────────────────────────────────────────────────────────────────────────────────────────
  Install:
───────────────────────────────────────────────────────────────────────────────────────────────────

  + abseil-cpp               20210324.2  h9c3ff4c_0              conda-forge/linux-64     1010 KB
  + absl-py                      0.15.0  pyhd8ed1ab_0            conda-forge/noarch         98 KB
  + aiohttp                       3.8.1  py39h3811e60_0          conda-forge/linux-64      588 KB
  + aiosignal                     1.2.0  pyhd8ed1ab_0            conda-forge/noarch         12 KB
  + astunparse                    1.6.3  pyhd8ed1ab_0            conda-forge/noarch         15 KB
  + async-timeout                 4.0.2  pyhd8ed1ab_0            conda-forge/noarch          9 KB
  + blinker                         1.4  py_1                    conda-forge/noarch        Cached
  + cached-property               1.5.2  hd8ed1ab_1              conda-forge/noarch          4 KB
  + cached_property               1.5.2  pyha770c72_1            conda-forge/noarch        Cached
  + cachetools                    4.2.4  pyhd8ed1ab_0            conda-forge/noarch         12 KB
  + click                         8.0.3  py39hf3d152e_1          conda-forge/linux-64      147 KB
  + cudatoolkit                  11.6.0  habf752d_10             conda-forge/linux-64      821 MB
  + cudnn                      8.2.1.32  h86fa8c9_0              conda-forge/linux-64      674 MB
  + frozenlist                    1.3.0  py39h3811e60_0          conda-forge/linux-64      161 KB
  + gast                          0.4.0  pyh9f0ad1d_0            conda-forge/noarch         12 KB
  + giflib                        5.2.1  h36c2ea0_2              conda-forge/linux-64      Cached
  + google-auth                  1.35.0  pyh6c4a22f_0            conda-forge/noarch         81 KB
  + google-auth-oauthlib          0.4.6  pyhd8ed1ab_0            conda-forge/noarch         19 KB
  + google-pasta                  0.2.0  pyh8c360ce_0            conda-forge/noarch         42 KB
  + grpc-cpp                     1.41.1  h75e9d12_2              conda-forge/linux-64        4 MB
  + grpcio                       1.41.1  py39hff7568b_1          conda-forge/linux-64        2 MB
  + h5py                          3.1.0  nompi_py39h25020de_100  conda-forge/linux-64        1 MB
  + hdf5                         1.10.6  nompi_h6a2412b_1114     conda-forge/linux-64        3 MB
  + jpeg                             9e  h7f98852_0              conda-forge/linux-64      Cached
  + keras                         2.6.0  pyhd8ed1ab_1            conda-forge/noarch        822 KB
  + keras-preprocessing           1.1.2  pyhd8ed1ab_0            conda-forge/noarch         34 KB
  + libblas                       3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libcblas                      3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libgfortran-ng               11.2.0  h69a702a_12             conda-forge/linux-64       19 KB
  + libgfortran5                 11.2.0  h5c6108e_12             conda-forge/linux-64      Cached
  + liblapack                     3.9.0  13_linux64_openblas     conda-forge/linux-64       12 KB
  + libopenblas                  0.3.18  pthreads_h8fe5266_0     conda-forge/linux-64      Cached
  + libpng                       1.6.37  h21135ba_2              conda-forge/linux-64      Cached
  + libprotobuf                  3.18.1  h780b84a_0              conda-forge/linux-64        3 MB
  + markdown                      3.3.6  pyhd8ed1ab_0            conda-forge/noarch         67 KB
  + multidict                     6.0.2  py39h3811e60_0          conda-forge/linux-64      156 KB
  + nccl                       2.11.4.1  h5c60f85_2              conda-forge/linux-64      146 MB
  + numpy                        1.19.5  py39hdbf815f_2          conda-forge/linux-64        5 MB
  + oauthlib                      3.2.0  pyhd8ed1ab_0            conda-forge/noarch         90 KB
  + opt_einsum                    3.3.0  pyhd8ed1ab_1            conda-forge/noarch        Cached
  + protobuf                     3.18.1  py39he80948d_0          conda-forge/linux-64      347 KB
  + pyasn1                        0.4.8  py_0                    conda-forge/noarch        Cached
  + pyasn1-modules                0.2.7  py_0                    conda-forge/noarch        Cached
  + pyjwt                         2.3.0  pyhd8ed1ab_1            conda-forge/noarch        Cached
  + python-flatbuffers             1.12  pyhd8ed1ab_1            conda-forge/noarch         19 KB
  + pyu2f                         0.1.5  pyhd8ed1ab_0            conda-forge/noarch        Cached
  + re2                      2021.11.01  h9c3ff4c_0              conda-forge/linux-64      222 KB
  + requests-oauthlib             1.3.1  pyhd8ed1ab_0            conda-forge/noarch         22 KB
  + rsa                             4.8  pyhd8ed1ab_0            conda-forge/noarch        Cached
  + scipy                         1.8.0  py39hee8e79c_1          conda-forge/linux-64       25 MB
  + snappy                        1.1.8  he1b5a44_3              conda-forge/linux-64       32 KB
  + tensorboard                   2.6.0  pyhd8ed1ab_1            conda-forge/noarch          5 MB
  + tensorboard-data-server       0.6.0  py39h95dcef6_1          conda-forge/linux-64        3 MB
  + tensorboard-plugin-wit        1.8.1  pyhd8ed1ab_0            conda-forge/noarch        668 KB
  + tensorflow                    2.6.2  cuda112py39h9333c2f_1   conda-forge/linux-64       29 KB
  + tensorflow-base               2.6.2  cuda112py39he9472f8_1   conda-forge/linux-64      331 MB
  + tensorflow-estimator          2.6.2  cuda112py39h9333c2f_1   conda-forge/linux-64      654 KB
  + termcolor                     1.1.0  py_2                    conda-forge/noarch        Cached
  + typing-extensions           3.7.4.3  0                       conda-forge/noarch          8 KB
  + typing_extensions           3.7.4.3  py_0                    conda-forge/noarch         25 KB
  + werkzeug                      2.0.3  pyhd8ed1ab_1            conda-forge/noarch        306 KB
  + wrapt                        1.12.1  py39h3811e60_3          conda-forge/linux-64       47 KB
  + yarl                          1.7.2  py39h3811e60_1          conda-forge/linux-64      138 KB

  Downgrade:
───────────────────────────────────────────────────────────────────────────────────────────────────

  - six                          1.16.0  pyh6c4a22f_0            installed                       
  + six                          1.15.0  pyh9f0ad1d_0            conda-forge/noarch         14 KB

  Summary:

  Install: 63 packages
  Downgrade: 1 packages

  Total download: 2 GB

───────────────────────────────────────────────────────────────────────────────────────────────────

Confirm changes: [Y/n] 
```
</details>

Reply:
Oops, wait a second...

Reply:
Okay, from testing locally, this doesn't seem to actually work. Here's what I gathered `CONDA_OVERRIDE_CUDA="x" mamba create --name tf tensorflow`
- it doesn't work with the intended consequence if x<10.2
- it works exactly the same for all x>=10.2 (e.g. x=10.2, x=11.2, x=11.6, x=116, x=500000) and it always downloads `cudatoolkit==10.2`

(Of course one can instruct it to get a specific `cudatoolkit`, or do something like `tensorflow=2.7.0=cuda112` --- this last one, `tensorflow=2.7.0=cuda112` naturally yields `cudatoolkit==11.6`)

I need to rewrite this section completely then.

Reply:
@isuruf could you check double check this? `CONDA_OVERRIDE_CUDA` (sorry about the typo in the writeup) seems buggy...

Reply:
@hmaarrfk could you correct me if I am wrong here? Could you check again?

Reply:
maybe related: https://github.com/conda/conda/issues/10803

Reply:
`CONDA_OVERRIDE_CUDA` allows overriding the *newest* supported CUDA version that conda allows. As long as it's above `10.2`, it will still happily allow installing 10.2.

Reply:
@kkraus14, what I am seeing is that it allows installing any cuda version, but defaults to 10.2 (i.e. setting `__cuda` to 10.2?)

Reply:
> `CONDA_OVERRIDE_CUDA` allows overriding the _newest_ supported CUDA toolkit version that conda allows. As long as it's above `10.2`, it will still happily allow installing 10.2.

I just checked this: `CONDA_OVERRIDE_CUDA` simply assigns `__cuda` whatever you give it... but I still don't understand why it is not working properly...

Reply:
(working properly = installing the specific cudatoolkit, if I am understanding how this should work fully)

Reply:
Oh --- are you saying that conda will always default to 10.2 regardless to what `__cuda` is at a given moment? (as long as it is 10.2+)

Reply:
```
CONDA_OVERRIDE_CUDA="x" mamba info

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.20.0) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


     active environment : None
            shell level : 0
       user config file : /home/ngam/.condarc
 populated config files : /home/ngam/.Mambaforge-Linux-x86_64/.condarc
                          /home/ngam/.condarc
          conda version : 4.11.0
    conda-build version : not installed
         python version : 3.9.10.final.0
       virtual packages : __cuda=x=0
                          __linux=5.4.0=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/ngam/.Mambaforge-Linux-x86_64  (writable)
      conda av data dir : /home/ngam/.Mambaforge-Linux-x86_64/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/ngam/.Mambaforge-Linux-x86_64/pkgs
                          /home/ngam/.conda/pkgs
       envs directories : /home/ngam/.Mambaforge-Linux-x86_64/envs
                          /home/ngam/.conda/envs
               platform : linux-64
             user-agent : conda/4.11.0 requests/2.27.1 CPython/3.9.10 Linux/5.4.0-97-generic ubuntu/20.04.3 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False

```

Reply:
What the `__cuda` virtual package does is if it finds the CUDA driver on the machine it calls the `cuDriverGetVersion` API (https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VERSION.html#group__CUDA__VERSION). This API returns the latest CUDA toolkit version supported by the driver. What this means is that any CUDA toolkit version older than this version will happily install.

This effectively enforces: `cudatoolkit <= __cuda` (with exceptions now surrounding cudatoolkit >= 11.1 because of CUDA enhanced compatibility).

Because of this, you can't control the version of tensorflow that will get installed using the override environment variable. That just allows the constraint to be resolved, but doesn't influence the version of of the toolkit further than ensuring compatibility. Why the solver is prioritizing the 10.2 package is unclear to me, but if you want the 11.2 package of tensorflow you should use:
```
CONDA_CUDA_OVERRIDE="11.2" mamba install tensorflow cudatoolkit=11.2 -c conda-forge
```

Reply:
Thank you so much for explaining this! 

Okay, at least I now understand the underlying mechanics and I think we are on the same page. 

I do wonder if constraining `__cuda` _with a version number_ in meta.yaml somehow helps here... 

At any rate, for the purpose of the documentation, I believe we should add `cudatoolkit=x` (as you suggest) as well just to alleviate any potential confusion with users ending up with 10.2 no matter what. 

So the above documentation part will read something like:

```
CONDA_CUDA_OVERRIDE="11.2" conda install tensorflow  cudatoolkit=11.2 -c conda-forge
...
Note that you should select the cudatoolkit version most appropraite for your GPU; for example, this can be "10.2", "11.0", "11.1", and "11.2" through "11.6". 
```

Does that sound like potentially a save approach? Should we maybe have "11.6" in the first line since it's the latest?

Reply:
> Note that you should select the cudatoolkit version most appropraite for your GPU; for example, this can be "10.2", "11.0", "11.1", and "11.2" through "11.6". 

This line is incorrect and should indicate that a version sufficiently new enough for your GPU should be selected. I.E. if you have an A100 you need at least CUDA 11.0, but if you have say a V100 you can still just as happily use CUDA 11.6. Now it also depends what driver is installed on the downstream system the environment is going to be used on.

Reply:
I thought "cudatoolkit version most appropraite for your GPU" would cover what you're saying, but let me rephrase it with the newfound understanding. I like "sufficiently new" ...

Reply:
Thanks @leofang @kkraus14 for jumping in & helping out here!

> Why the solver is prioritizing the 10.2 package is unclear to me, but if you want the 11.2 package of tensorflow you should use:
  ```
  CONDA_CUDA_OVERRIDE="11.2" mamba install tensorflow cudatoolkit=11.2 -c conda-forge
  ```

To be honest, that sounds like a bug to me, especially if a cudatoolkit is already installed - specifying `cudatoolkit=11.2` in addition to `CONDA_CUDA_OVERRIDE="11.2"` feels redundant and brittle - the override itself should suffice to choose a version, no?

Reply:
Do you think it would work to add a version constraint to the `__cuda` run-requirement à la:
```
- __cuda >={{ cuda_compiler_version }}  # [cuda_compiler_version != "None"]
```

cudatoolkit itself does something [similar](https://github.com/conda-forge/cudatoolkit-feedstock/blob/master/recipe/meta.yaml#L525-L530), but not as a run-export.

Reply:
> the override itself should suffice to choose a version, no?

No, the override is purely to indicate that the driver should support up to CUDA Toolkit 11.2. Using any version of the CUDA Toolkit less than or equal to 11.2 is valid. Assuming all other things are equal the conda / mamba solver should prioritize getting the newest `cudatoolkit` version possible, but it seems like there's additional things influencing as to why it's solving to 10.2 instead.

Reply:
OK, thanks for the clarification!

> Assuming all other things are equal the conda / mamba solver should prioritize getting the newest `cudatoolkit` version possible, but it seems like there's additional things influencing as to why it's solving to 10.2 instead.

Well if it _is_ solvable for `cudatoolkit=11.2`, then not getting it after passing 11.2 to the override explicitly is perhaps not a bug but at least a bad user experience (and hard to explain, as we're seeing in this discussion). My point with the constraint was if we can provide enough crutches to the solver that it selects the override (or newer for the CEH ones).

Reply:
> For example, I could have a project that uses both the runtime API via the CUDA Toolkit and the CUDA Driver API directly and requires Toolkit 10.2 and Driver version >=450 which corresponds to `__cuda>=11.0`. If someone wanted to install this on a machine without GPUs then it would look like `CONDA_CUDA_OVERRIDE="11.0" mamba install my_package cudatoolkit=10.2`

This really forms a much better picture in my mind now, thank you so much!! I would like to test if this is achievable, so I will look into it further :)

Reply:
[resolving this; let me know if I still miss the point in the edits]